File: grind75/week2/010-Partition-Equal-Subset-Sum-Medium/010-Partition-Equal-Subset-Sum-Medium.py
# Good solution, not the most optimal
# Most optimal is 1-D DP array
class Solution:
    def canPartition(self, nums: List[int]) -> bool:

        # base case check
        # sum of nums divided by 2 must be an even number, otherwise you can't have two equal partitions
        # such as if the sum of the input is like 7 or something, then we cant obviously have two equal
        # partitions
        total = sum(nums)
        if total % 2 != 0:
            return False

        # Each partition must be equal to the total sum of the nums array divided by 2
        # E.g., for [1,5,11,5] => 1 + 5 + 11 + 5 = 22, so 22 / 2 = 11
        # So each partition must sum up to 11
        target = total // 2
        
        # to build up to target, we either take or don't take, like knapsack
        # ------
        # duplicates wont get stored in the set
        dp = set()
        dp.add(0) # base case

        for i in range(len(nums)):
            # Cannot update dp set while iterating over it, so use a separate set, nextDP
            nextDP = set()
            for t in dp:
                nextDP.add(t + nums[i])
                nextDP.add(t) 
            dp = nextDP

        return True if target in dp else False


# Optimal 1-D DP Approach!

class Solution:
    def canPartition(self, nums: List[int]) -> bool:

        total = sum(nums)

        if total % 2 != 0:
            return False
        
        target = total // 2

        # dp[i] = True if we can form sum i using some subset of nums
        dp = [False] * (target + 1)
        dp[0] = True # base case: sum of 0 is always possible with no elements
        
        for num in nums:
            for j in range(target, num - 1, -1):
                if dp[j - num]:
                    dp[j] = True
        

        return dp[target]

        

File: grind75/week2/010-Partition-Equal-Subset-Sum-Medium/010-Partition-Equal-Subset-Sum-Medium.cpp
// Problem link: https://leetcode.com/problems/partition-equal-subset-sum/

// Not the most optimal, but an elegant solution nonetheless using a set
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        // Base case check
        // Sum of nums must be even, otherwise you can't split into two equal partitions
        // For example, if the sum is 7, it's impossible to divide it into two equal parts
        int total = 0;
        for (int num : nums) 
        {
            total += num;
        }
        if (total % 2 != 0) 
        {
            return false;
        }

        // Each partition must sum to total / 2
        // E.g., for [1,5,11,5] => 1 + 5 + 11 + 5 = 22, so target = 11
        int target = total / 2;

        // To build up to the target, we either take or don't take each number — like knapsack
        // ------
        // Duplicates won't get stored in an unordered_set
        unordered_set<int> dp;
        dp.insert(0);  // Base case: sum of 0 is always possible

        for (int i = 0; i < nums.size(); i++) 
        {
            // Cannot update dp set while iterating over it, so use a separate set: nextDP
            unordered_set<int> nextDP;
            for (int t : dp) 
            {
                nextDP.insert(t + nums[i]);
                nextDP.insert(t);
            }
            dp = nextDP;
        }
        print("Possible subset sums:", dp)
        // Final check: was the target sum reachable?
        return dp.count(target) > 0;
    }
};

// Optimal 1-D DP approach

class Solution {
public:
    bool canPartition(vector<int>& nums) {
        // Calculate the total sum of the array
        int total = 0;
        for (int num : nums) 
        {
            total += num;
        }

        // If the total sum is odd, we can't split it into two equal subsets
        if (total % 2 != 0) 
        {
            return false;
        }

        // Each partition must sum to total / 2
        int target = total / 2;

        // dp[i] = true means we can form sum i using some subset of nums
        vector<bool> dp(target + 1, false);
        dp[0] = true; // Base case: sum 0 is always possible with no elements

        // Iterate through each number in the input array
        for (int num : nums) 
        {
            // Loop backwards to avoid using the same number more than once in this round
            for (int j = target; j >= num; j--) 
            {
                if (dp[j - num]) 
                {
                    dp[j] = true;
                }
            }
        }

        // Final check: can we form the target sum?
        return dp[target];
    }
};
